### PR TITLE
podman: add nvidia runtime support

### PR DESCRIPTION
**Pull request by cpcloud**
###### Motivation for this change

This change adds an `enableNvidia` option to the `podman` module that enables the nvidia runtime.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by zowoq** [comment]
Please either undo the formatting changes or move them into a separate commit.

File: nixos/modules/virtualisation/podman.nix [hunk 1]
```diff
@@ -74,25 +82,50 @@ in
...
+          message = "Option dockerCompat conflicts with docker";
+        }
+        {
+          assertion = cfg.enableNvidia -> !config.virtualisation.docker.enableNvidia;
+          message = "Option enableNvidia conflicts with docker.enableNvidia";
+        }
+      ];
+    }
+    (mkIf cfg.enableNvidia {
+      environment.etc."nvidia-container-runtime/config.toml".source =
```
Comment (zowoq) [inline nixos/modules/virtualisation/podman.nix:118]: The docker version of this lives in the package and the docker module only needs to source it, I think we should to the same here.

**Comment by zowoq** [comment]
Sorry, I meant format then apply your changes. Add things and then reformatting them is hard to follow.

File: nixos/modules/virtualisation/podman.nix (hunk 1 above)
Comment (cpcloud) [reply nixos/modules/virtualisation/podman.nix:118]: There are some values that are different between the podman version and docker version though: https://github.com/NVIDIA/nvidia-container-runtime/issues/85#issuecomment-604931556. How should we handle that?

Comment (cpcloud) [reply nixos/modules/virtualisation/podman.nix:118]: In particular `no-cgroups = true` for podman (rootless) and `no-cgroups = false` for docker and podman rootful.

Comment (zowoq) [reply nixos/modules/virtualisation/podman.nix:118]: Create `pkgs/applications/virtualization/nvidia-docker/podman-config.toml` and source that here.

Comment (cpcloud) [reply nixos/modules/virtualisation/podman.nix:118]: The other issue is filling in the value of `glibc` which is done by filling in a the `installPhase` of `nvidia-docker`. That was the other reason I opted for the expression based version of this file.

**Comment by cpcloud** [comment]
Alright, I will redo.

File: nixos/modules/virtualisation/podman.nix (hunk 1 above)
Comment (zowoq) [reply nixos/modules/virtualisation/podman.nix:118]: Reusing the `@@glibcbin@` `substituteInPlace` should be fine.

File: nixos/modules/virtualisation/podman.nix [hunk 2]
```diff
@@ -120,16 +120,11 @@ in
...
-            ldcache = "/tmp/ld.so.cache";
-            load-kmods = true;
-            ldconfig = "@${lib.getBin glibc}/bin/ldconfig";
-            no-cgroups = true;
-          };
-        };
+        pkgs.runCommandNoCC "" { } ''
+          cp ${../../../pkgs/applications/virtualization/nvidia-docker/podman-config.toml} \
+            $out/config.toml
+          substituteInPlace $out/config.toml --subst-by-var glibcbin ${lib.getBin pkgs.glibc}
```
Comment (zowoq) [inline nixos/modules/virtualisation/podman.nix:126]: Move this to `nvidia-docker` with the docker `cp` and `substituteInPlace` then we should only need to source it here like the docker module does.

https://github.com/NixOS/nixpkgs/blob/e7fe577d9803885d1191c6612b95c246cb605dde/pkgs/applications/virtualization/nvidia-docker/default.nix#L62-L63

https://github.com/NixOS/nixpkgs/blob/9a63b3d3d61ca822aceeb4a994e8ef21a7593cc7/nixos/modules/virtualisation/docker.nix#L218-L220

Comment (cpcloud) [reply nixos/modules/virtualisation/podman.nix]: Done. Thanks for your patience.

**Comment by cpcloud** [comment]
@zowoq In general is it possible to test basic GPU functionality on the nix build farm? If so, I'll add a test here.

**Comment by zowoq** [comment]
> In general is it possible to test basic GPU functionality on the nix build farm?

No, I don't think we're able to test this.

Now that the formatting changes are split out from the other changes I've noticed that you're probably using `nixfmt`, this file is already almost compliant with `nixpkgs-fmt`, it would only need one change.

Rather than ask you to redo this again I've reverted most of the formatting changes, squashed commits as needed and updated the commit messages to [follow our style](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes).

**Comment by cpcloud** [comment]
Alright I will try to get my editor set up to use the right tool to auto format

**Comment by cpcloud** [comment]
I am using `nixpkgs-fmt` actually, I suppose the default config is aggressive/unstable?

**Comment by Mic92** [comment]
> I am using `nixpkgs-fmt` actually, I suppose the default config is aggressive/unstable?

It is. Especially nixpkgs is not aligned with it. Maybe this change in future.

**Comment by zowoq** [comment]
For me with `nixpkgs-fmt 1.0.0` formatting `nixos/modules/virtualisation/podman.nix` on master only makes this change:
```diff
diff --git a/nixos/modules/virtualisation/podman.nix b/nixos/modules/virtualisation/podman.nix
index f554aeffb45..e108d9d8c5f 100644
--- a/nixos/modules/virtualisation/podman.nix
+++ b/nixos/modules/virtualisation/podman.nix
@@ -7,10 +7,11 @@ let
   podmanPackage = (pkgs.podman.override { inherit (cfg) extraPackages; });

   # Provides a fake "docker" binary mapping to podman
-  dockerCompat = pkgs.runCommandNoCC "${podmanPackage.pname}-docker-compat-${podmanPackage.version}" {
-    outputs = [ "out" "man" ];
-    inherit (podmanPackage) meta;
-  } ''
+  dockerCompat = pkgs.runCommandNoCC "${podmanPackage.pname}-docker-compat-${podmanPackage.version}"
+    {
+      outputs = [ "out" "man" ];
+      inherit (podmanPackage) meta;
+    } ''
     mkdir -p $out/bin
     ln -s ${podmanPackage}/bin/podman $out/bin/docker

```

**Comment by cpcloud** [comment]
I am going to rebase this and make sure this works with https://github.com/NixOS/nixpkgs/pull/108607

File: pkgs/applications/virtualization/nvidia-docker/podman-config.toml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+[nvidia-container-cli]
+#root = "/run/nvidia/driver"
+#path = "/usr/bin/nvidia-container-cli"
+environment = []
+#debug = "/var/log/nvidia-container-runtime-hook.log"
+ldcache = "/tmp/ld.so.cache"
+load-kmods = true
+no-cgroups = true
+#user = "root:video"
+ldconfig = "@@glibcbin@/bin/ldconfig"
```
Comment (Mic92) [inline pkgs/applications/virtualization/nvidia-docker/podman-config.toml:13]: Zwei @? Is this correct?

Comment (cpcloud) [reply pkgs/applications/virtualization/nvidia-docker/podman-config.toml:13]: It is. Let me dig around for reference material.

Comment (cpcloud) [reply pkgs/applications/virtualization/nvidia-docker/podman-config.toml:13]: The `@glibcbin@` is used for templating in that value using `substituteInPlace` while the first `@` indicates to `libnvidia-container` that the path to `ldconfig` is relative to the host (https://github.com/NVIDIA/libnvidia-container/blob/master/src/nvc_ldcache.c#L350-L359).

**Comment by cpcloud** [comment]
@zowoq I ended up having to add `runc` because that's being invoked by the new `nvidia-container-runtime` update I made in another PR. This could use another review.

File: nixos/modules/virtualisation/podman.nix [hunk 4]
```diff
@@ -78,21 +88,37 @@ in
...
+          }));
+      };
 
+      assertions = [
+        {
+          assertion = cfg.dockerCompat -> !config.virtualisation.docker.enable;
+          message = "Option dockerCompat conflicts with docker";
+        }
+        {
+          assertion = cfg.enableNvidia -> !config.virtualisation.docker.enableNvidia;
```
Comment (cpcloud) [inline nixos/modules/virtualisation/podman.nix:115]: I plan to remove this restriction in a follow up PR.

**Comment by cpcloud** [comment]
@Mic92 @zowoq Can one of y'all give this a final review (and/or merge if satisifed) at your earliest convenience? Thanks!
